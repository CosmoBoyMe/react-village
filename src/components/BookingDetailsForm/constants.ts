const DAYS_DECLENSIONS = ['сутки', 'суток', 'суток'];
const BOOKING_FORM_TOAST_ID = 'BOOKING_FORM_TOAST_ID';
const GUESTS_DECLENSIONS = {
  adults: 'взрослых',
  children: 'детей',
  babies: 'младенцев',
};

export { BOOKING_FORM_TOAST_ID, DAYS_DECLENSIONS, GUESTS_DECLENSIONS };
