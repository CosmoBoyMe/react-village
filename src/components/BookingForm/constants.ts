const DAYS_DECLINATIONS = ['сутки', 'суток', 'суток'];

export { DAYS_DECLINATIONS };
