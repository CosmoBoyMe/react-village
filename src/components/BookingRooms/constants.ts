const TABS_BUTTONS_DATA = [
  { name: 'все' },
  { name: 'текущие' },
  { name: 'прошедшие' },
];

export { TABS_BUTTONS_DATA };
