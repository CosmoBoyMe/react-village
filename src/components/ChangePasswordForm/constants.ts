const enum ChangePasswordFormNames {
  Password = 'password',
  NewPassword = 'newPassword',
  PasswordConfirm = 'passwordConfirm',
}

const CHANGE_PASSWORD_FORM_TOAST_ID = 'CHANGE_PASSWORD_FORM_TOAST_ID';

export { CHANGE_PASSWORD_FORM_TOAST_ID, ChangePasswordFormNames };
