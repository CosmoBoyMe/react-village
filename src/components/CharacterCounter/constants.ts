const RADIUS = 20;

export { RADIUS };
