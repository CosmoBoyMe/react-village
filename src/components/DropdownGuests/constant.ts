const DROPDOWN_GUESTS_DECLENSIONS = {
  guests: ['гость', 'гостя', 'гостей'],
  babies: ['младенец', 'младенца', 'младенцев'],
};

export { DROPDOWN_GUESTS_DECLENSIONS };
