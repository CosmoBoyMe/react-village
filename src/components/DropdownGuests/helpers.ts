import { DropdownGuestsIds } from '../../shared/constants/DropdownGuestsIds';
import { declination } from '../../shared/helpers/declination/declination';

export type DropdownGuestsItemData = {
  id: DropdownGuestsIds;
  name: string;
  amount: number;
};

export type DropdownItemsDeclensions = {
  guests: string[];
  babies: string[];
};

const getCorrectDropdownValue = (
  items: DropdownGuestsItemData[],
  declensions: DropdownItemsDeclensions,
  totalGuests: number
) => {
  const value = [];

  const babiesAmount = Number(
    items.find((item) => item.id === 'babies')?.amount
  );

  if (totalGuests > 0) {
    const guestsValue = `${totalGuests} ${declination(
      totalGuests,
      declensions.guests
    )}`;

    value.push(guestsValue);
  }

  if (babiesAmount > 0) {
    const babiesValue = `${babiesAmount} ${declination(
      babiesAmount,
      declensions.babies
    )}`;

    value.push(babiesValue);
  }

  return value;
};

export { getCorrectDropdownValue };
