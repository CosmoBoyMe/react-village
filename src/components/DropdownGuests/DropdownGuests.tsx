import { FC, useEffect, useRef, useState } from 'react';
import classNames from 'classnames';

import { getUniqueArray } from '../../shared/helpers/getUniqueArray/getUniqueArray';
import { DropdownGuestsItemData } from '../../types/DropdownItemData';
import { Button } from '../Button/Button';

import { DropdownGuestsItem } from './DropdownGuestsItem/DropdownGuestsItem';
import { DROPDOWN_GUESTS_DECLENSIONS } from './constant';
import { getCorrectDropdownGuestsValue } from './helpers';
import '../../styles/Dropdown.scss';

type Props = {
  items: DropdownGuestsItemData[];
  guestsLimit?: number;
  babiesLimit?: number;
  onChange?: (dropdownItems: DropdownGuestsItemData[]) => void;
};

const DropdownGuests: FC<Props> = ({
  items,
  guestsLimit = 15,
  babiesLimit = 10,
  onChange,
}) => {
  const dropdownRef = useRef<HTMLDivElement>(null);
  const [isOpen, setIsOpen] = useState(false);
  const dropdownItems = getUniqueArray(items, 'id');

  const totalAmount = dropdownItems.reduce((acc, item) => acc + item.amount, 0);

  const guestsAmount = dropdownItems
    .filter((item) => item.id !== 'babies')
    .reduce((acc, item) => acc + item.amount, 0);

  const babiesAmount = Number(
    dropdownItems.find((item) => item.id === 'babies')?.amount
  );
  const adultsAmount = Number(
    dropdownItems.find((item) => item.id === 'adults')?.amount
  );

  const setBabiesToZero = (currentItems: DropdownGuestsItemData[]) => {
    const newItems = currentItems.map((item) => {
      if (item.id === 'babies') {
        return {
          ...item,
          amount: 0,
        };
      }

      return item;
    });

    onChange?.(newItems);
  };

  const hasCurrentItemMaxValue = (id: string) => {
    if (id === 'babies') {
      return adultsAmount <= 0 || babiesAmount >= babiesLimit;
    }

    return guestsAmount >= guestsLimit;
  };

  const clear = () => {
    const newItems = dropdownItems.map((item) => {
      return {
        ...item,
        amount: 0,
      };
    });

    onChange?.(newItems);
  };

  const handleCounterChange = (name: string, amount: number) => {
    const newItems = dropdownItems.map((item) => {
      let currentItem = item;

      if (item.name === name) {
        currentItem = {
          ...item,
          amount,
        };
      }

      return currentItem;
    });

    const currentAdultsAmount = Number(
      newItems.find((item) => item.id === 'adults')?.amount
    );

    if (currentAdultsAmount <= 0) {
      setBabiesToZero(newItems);
      return;
    }

    onChange?.(newItems);
  };

  useEffect(() => {
    const handleDocumentPointerDown = ({ target }: PointerEvent) => {
      const dropdown = dropdownRef.current;
      const dropdownContainsTarget =
        target instanceof Node && dropdown && dropdown.contains(target);

      if (!dropdownContainsTarget) {
        setIsOpen(false);
      }
    };

    document.addEventListener('pointerdown', handleDocumentPointerDown);

    return () =>
      document.removeEventListener('pointerdown', handleDocumentPointerDown);
  }, []);

  const handleDropdownPointerDown = () => {
    setIsOpen((prevState) => !prevState);
  };

  const handleClearButtonPointerDown = () => {
    clear();
  };

  const handleApplyButtonPointerDown = () => {
    setIsOpen(false);
  };

  return (
    <div
      data-testid="guest-dropdown"
      ref={dropdownRef}
      className={classNames('dropdown', {
        dropdown_opened: isOpen,
      })}
    >
      <h3 className="dropdown__heading">Гости</h3>
      <div className="dropdown__wrapper">
        <input
          className="dropdown__input"
          type="text"
          placeholder="Сколько гостей"
          value={getCorrectDropdownGuestsValue(
            DROPDOWN_GUESTS_DECLENSIONS,
            guestsAmount,
            babiesAmount
          ).join(', ')}
          onClick={handleDropdownPointerDown}
          readOnly
        />
        <button
          className={classNames('dropdown__arrow-button', 'material-icons', {
            'dropdown__arrow-button_rotate': isOpen,
          })}
          type="button"
          onClick={handleDropdownPointerDown}
        >
          expand_more
        </button>
        <div className="dropdown__drop">
          <ul className="dropdown__list">
            {dropdownItems.map(({ id, name, amount }) => (
              <DropdownGuestsItem
                key={id}
                name={name}
                amount={amount}
                onChangeCounter={handleCounterChange}
                incrementDisabled={hasCurrentItemMaxValue(id)}
              />
            ))}
          </ul>
          <div className="dropdown__buttons">
            <div
              className={classNames('dropdown__button-clear', {
                dropdown__button_hidden: totalAmount <= 0,
              })}
            >
              <Button text="Очистить" onClick={handleClearButtonPointerDown} />
            </div>
            <div className="dropdown__button-apply">
              <Button text="Применить" onClick={handleApplyButtonPointerDown} />
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export { DropdownGuests };
