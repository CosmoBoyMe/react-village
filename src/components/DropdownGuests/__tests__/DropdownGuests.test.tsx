/* eslint-disable @typescript-eslint/await-thenable */
/* eslint-disable testing-library/render-result-naming-convention */

import { fireEvent, render, screen } from '@testing-library/react';

import '@testing-library/jest-dom';

import { DropdownGuestsIds } from '../../../shared/constants/DropdownGuestsIds';
import { DropdownGuestsItemData } from '../../../types/DropdownItemData';
import { DropdownGuests } from '../DropdownGuests';

describe('dropdownGuests component rendering', () => {
  it(`Return item value when Plus or Minus button clicked`, () => {
    const guests = [
      { id: DropdownGuestsIds.ADULTS, name: 'взрослые', amount: 0 },
      { id: DropdownGuestsIds.CHILDREN, name: 'дети', amount: 0 },
      { id: DropdownGuestsIds.BABIES, name: 'младенцы', amount: 0 },
    ];

    const onChange = jest.fn(
      (dropdownItems: DropdownGuestsItemData[]) => dropdownItems
    );
    const dropdownGuests = render(
      <DropdownGuests items={guests} onChange={onChange} />
    );

    expect(dropdownGuests).toMatchSnapshot();

    const applyButton = screen.getByText('Применить');

    const buttonsPlus = screen.getAllByText('+');
    expect(buttonsPlus).toHaveLength(3);
    const buttonsMinus = screen.getAllByText('-');
    fireEvent.click(buttonsPlus[0]);
    expect(onChange).lastReturnedWith([
      { id: 'adults', name: 'взрослые', amount: 1 },
      { id: 'children', name: 'дети', amount: 0 },
      { id: 'babies', name: 'младенцы', amount: 0 },
    ]);

    fireEvent.click(buttonsPlus[1]);
    expect(onChange).lastReturnedWith([
      { id: 'adults', name: 'взрослые', amount: 1 },
      { id: 'children', name: 'дети', amount: 1 },
      { id: 'babies', name: 'младенцы', amount: 0 },
    ]);
    fireEvent.click(buttonsPlus[2]);
    expect(onChange).lastReturnedWith([
      { id: 'adults', name: 'взрослые', amount: 1 },
      { id: 'children', name: 'дети', amount: 1 },
      { id: 'babies', name: 'младенцы', amount: 1 },
    ]);
    fireEvent.click(buttonsMinus[0]);
    expect(onChange).lastReturnedWith([
      { id: 'adults', name: 'взрослые', amount: 0 },
      { id: 'children', name: 'дети', amount: 1 },
      { id: 'babies', name: 'младенцы', amount: 0 },
    ]);
    fireEvent.click(buttonsMinus[1]);
    expect(onChange).lastReturnedWith([
      { id: 'adults', name: 'взрослые', amount: 0 },
      { id: 'children', name: 'дети', amount: 0 },
      { id: 'babies', name: 'младенцы', amount: 0 },
    ]);

    fireEvent.click(applyButton);
    expect(dropdownGuests).toMatchSnapshot();
  });
});
