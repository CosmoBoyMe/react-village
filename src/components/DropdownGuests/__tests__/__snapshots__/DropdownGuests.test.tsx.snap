// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdownGuests component rendering dropdownGuests layout matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="dropdown"
        data-testid="guest-dropdown"
      >
        <h3
          class="dropdown__heading"
        >
          Гости
        </h3>
        <div
          class="dropdown__wrapper"
        >
          <input
            class="dropdown__input"
            placeholder="Сколько гостей"
            readonly=""
            type="text"
            value=""
          />
          <button
            class="dropdown__arrow-button material-icons"
            type="button"
          >
            expand_more
          </button>
          <div
            class="dropdown__drop"
          >
            <ul
              class="dropdown__list"
            >
              <li
                class="dropdown-item"
              >
                <p
                  class="dropdown-item__name"
                >
                  взрослые
                </p>
                <div
                  class="dropdown-item__controls"
                >
                  <button
                    class="dropdown-item__button"
                    disabled=""
                    type="button"
                  >
                    -
                  </button>
                  <p
                    class="dropdown-item__amount"
                    data-testid="dropdown-guest-amount"
                  >
                    0
                  </p>
                  <button
                    class="dropdown-item__button"
                    type="button"
                  >
                    +
                  </button>
                </div>
              </li>
              <li
                class="dropdown-item"
              >
                <p
                  class="dropdown-item__name"
                >
                  дети
                </p>
                <div
                  class="dropdown-item__controls"
                >
                  <button
                    class="dropdown-item__button"
                    disabled=""
                    type="button"
                  >
                    -
                  </button>
                  <p
                    class="dropdown-item__amount"
                    data-testid="dropdown-guest-amount"
                  >
                    0
                  </p>
                  <button
                    class="dropdown-item__button"
                    type="button"
                  >
                    +
                  </button>
                </div>
              </li>
              <li
                class="dropdown-item"
              >
                <p
                  class="dropdown-item__name"
                >
                  младенцы
                </p>
                <div
                  class="dropdown-item__controls"
                >
                  <button
                    class="dropdown-item__button"
                    disabled=""
                    type="button"
                  >
                    -
                  </button>
                  <p
                    class="dropdown-item__amount"
                    data-testid="dropdown-guest-amount"
                  >
                    0
                  </p>
                  <button
                    class="dropdown-item__button"
                    disabled=""
                    type="button"
                  >
                    +
                  </button>
                </div>
              </li>
            </ul>
            <div
              class="dropdown__buttons"
            >
              <div
                class="dropdown__button-clear dropdown__button_hidden"
              >
                <button
                  class="button"
                  type="button"
                >
                  Очистить
                </button>
              </div>
              <div
                class="dropdown__button-apply"
              >
                <button
                  class="button"
                  type="button"
                >
                  Применить
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="dropdown"
      data-testid="guest-dropdown"
    >
      <h3
        class="dropdown__heading"
      >
        Гости
      </h3>
      <div
        class="dropdown__wrapper"
      >
        <input
          class="dropdown__input"
          placeholder="Сколько гостей"
          readonly=""
          type="text"
          value=""
        />
        <button
          class="dropdown__arrow-button material-icons"
          type="button"
        >
          expand_more
        </button>
        <div
          class="dropdown__drop"
        >
          <ul
            class="dropdown__list"
          >
            <li
              class="dropdown-item"
            >
              <p
                class="dropdown-item__name"
              >
                взрослые
              </p>
              <div
                class="dropdown-item__controls"
              >
                <button
                  class="dropdown-item__button"
                  disabled=""
                  type="button"
                >
                  -
                </button>
                <p
                  class="dropdown-item__amount"
                  data-testid="dropdown-guest-amount"
                >
                  0
                </p>
                <button
                  class="dropdown-item__button"
                  type="button"
                >
                  +
                </button>
              </div>
            </li>
            <li
              class="dropdown-item"
            >
              <p
                class="dropdown-item__name"
              >
                дети
              </p>
              <div
                class="dropdown-item__controls"
              >
                <button
                  class="dropdown-item__button"
                  disabled=""
                  type="button"
                >
                  -
                </button>
                <p
                  class="dropdown-item__amount"
                  data-testid="dropdown-guest-amount"
                >
                  0
                </p>
                <button
                  class="dropdown-item__button"
                  type="button"
                >
                  +
                </button>
              </div>
            </li>
            <li
              class="dropdown-item"
            >
              <p
                class="dropdown-item__name"
              >
                младенцы
              </p>
              <div
                class="dropdown-item__controls"
              >
                <button
                  class="dropdown-item__button"
                  disabled=""
                  type="button"
                >
                  -
                </button>
                <p
                  class="dropdown-item__amount"
                  data-testid="dropdown-guest-amount"
                >
                  0
                </p>
                <button
                  class="dropdown-item__button"
                  disabled=""
                  type="button"
                >
                  +
                </button>
              </div>
            </li>
          </ul>
          <div
            class="dropdown__buttons"
          >
            <div
              class="dropdown__button-clear dropdown__button_hidden"
            >
              <button
                class="button"
                type="button"
              >
                Очистить
              </button>
            </div>
            <div
              class="dropdown__button-apply"
            >
              <button
                class="button"
                type="button"
              >
                Применить
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
